matron/src/screen_events.c

```diff
@@ -0,0 +1,199 @@
+#include <assert.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "screen_events.h"
+#include "screen.h"
+
+#define SCREEN_Q_SIZE 1024
+#define SCREEN_Q_MASK (SCREEN_Q_SIZE -1)
+
+static screen_event_data_t screen_q[SCREEN_Q_SIZE];
+static int screen_q_wr = 0;
+static int screen_q_rd = 0;
+
+static void* screen_event_loop(void*);
+static void screen_event_pop(screen_event_data_t *dst);
+static void handle_screen_event(const screen_event_data_t *data);
+
+static pthread_t screen_event_thread;
+static pthread_mutex_t screen_q_lock;
+static pthread_cond_t screen_q_nonempty;
+
+void screen_events_init() { 
+    pthread_cond_init(&screen_q_nonempty, NULL);
+    
+    // set up screen event loop
+    if (pthread_create(&screen_event_thread, NULL, screen_event_loop, 0) ) {
+      fprintf(stderr, "SCREEN: error creating thread\n");
+    }
+}
+
+void screen_event_push(screen_event_id_t type, const char *text, int data_count, ...) {
+    va_list args;
+    
+    pthread_mutex_lock(&screen_q_lock);
+    screen_event_data_t* data = &(screen_q[screen_q_wr]); 
+    data->type = type;
+    data->data_count = data_count;
+    if (text != NULL) {
+        int nbytes = strlen(text) + 1;
+        data->text = malloc(nbytes);
+        memcpy(data->text, text, nbytes);
+    } else {
+        data->text = NULL;
+    }
+    va_start(args, data_count);
+    for (int i=0; i<data_count; ++i) { 

```

Comment:
recommend bounds checking `data_count` against the max number of `data` values accommodated in the event struct (currently 6). in my book I’d probably just use an assert at the top of the function.